derivative for wigner, example of equivariance check, increase generality

bkmi opened the pull request:
Hopefully you like it!

bkmi committed:
added spherical harmonic parity operator.

bkmi committed:
added arbitrary parity operator.

bkmi committed:
changed default parity operators to float64

bkmi committed:
Merge branch 'point' of https://github.com/mariogeiger/se3cnn into point

bkmi committed:
removed parity, added equivariance test

bkmi committed:
Merge branch 'point' of https://github.com/mariogeiger/se3cnn into point

bkmi committed:
improved equivaraince check

bkmi committed:
Merge branch 'point' of https://github.com/mariogeiger/se3cnn into point

bkmi committed:
working on check network equivariance

bkmi committed:
removed equivariance check

bkmi committed:
remove bad stuff

bkmi committed:
added example.

mariogeiger commented:
Isn't `ph.derivative_rot_mat` returning 3 matrices? The derivative wrt alpha, wrt beta and wrt gamma?
if so, don't you also want to return 3 tensors?

bkmi committed:
Merge branch 'point' of https://github.com/mariogeiger/se3cnn into point

bkmi committed:
changed derivative to return a 3-tuple of tensors

bkmi commented:
I updated it to return a 3-tuple of tensors: dDda, dDdb, dDdc.
Does that solve it?

bkmi commented:
There is also the equivariance check. Perhaps we should wait to merge until we update the equivariance check readme.

bkmi committed:
rotation equivariance check, polymorphic checking of cg calculations

bkmi commented:
I added the equivariance check for the rotation case. Still need to do the parity one.

Note I changed the multiplication file because when using float32, the precision is too low to satisfy the `error < 1e-10` assertion.

mariogeiger commented:
Thanks Ben, shall I wait the parity test before merge?

bkmi committed:
added dtype and device to spherical harmonics xyz

bkmi committed:
fixed grad in forward kernel. Adjusted spherical harmonics xyz

bkmi commented:
Added a fix for #26.
Changed how spherical_harmonics_xyz is calculated.

bkmi commented:
> Thanks Ben, shall I wait the parity test before merge?

Hmm not sure. I am feeling pretty lazy about writing that right now, but I think there are other good fixes in here. I should have separated them.

bkmi committed:
removed a bad class

mariogeiger commented:
The `.cpu()` is not necessary.
About the `.detach()` I don’t think we should put it here. 
It’s an informative error for the user that the gradient cannot be computer. 
Moreover if the user changes the default sh by a backwardable one then we don’t want the `.detach()`. 
As a user you can do what you did by doing `sh=lambda x: sh(x.detach())` as argument of Kernel.

bkmi committed:
Added error for backwardable inputs to kernel.

bkmi committed:
better raise

bkmi committed:
added parity equivariance check

bkmi commented:
Alright. I vote that it's ready if you think it is. Please accept the pull or let me know what I should change.

mariogeiger reviewed (changes requested):
for me it looks like `examples/point/equivariance.py` should be put into the test files and use the same format as the other files in this directory.

mariogeiger commented on se3cnn/point/kernel.py:
```diff
@@ -113,7 +113,10 @@ def forward(self, r):
         # precompute all needed spherical harmonics
-        Ys = self.sh(self.set_of_l_filters, r)  # [l_filter * m_filter, batch]
+        if r.requires_grad:
+            raise TypeError("The r input of kernel cannot require a gradient.")
```
You cannot do that. Someone might want to allow `r` to require grad if he uses its own `sh` function.
Why do you want to do that? Is it the current error message that is not clear enough?

bkmi replied:
It was for clarity, but I'm okay with leaving it how it is:

```python
Traceback (most recent call last):
  File "/Users/user/miniconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3326, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-c6f6414c7b04>", line 1, in <module>
    runfile('/Users/user/sci/cgae/otp_equi.py', wdir='/Users/user/sci/cgae')
  File "/Applications/PyCharm.app/Contents/helpers/pydev/_pydev_bundle/pydev_umd.py", line 197, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "/Applications/PyCharm.app/Contents/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/Users/user/sci/cgae/otp_equi.py", line 223, in <module>
    main()
  File "/Users/user/sci/cgae/otp_equi.py", line 215, in main
    results = execute(args)
  File "/Users/user/sci/cgae/otp_equi.py", line 114, in execute
    pred_sph = decoder(cg_features, cg_xyz)
  File "/Users/user/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/Users/user/sci/cgae/cgae/equi.py", line 89, in forward
    output = layer(output.div(geometry.size(1) ** 0.5), geometry)  # Normalization of layers by number of atoms.
  File "/Users/user/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/Users/user/sci/se3cnn/se3cnn/non_linearities/gated_block.py", line 41, in forward
    features = self.op(*args, **kwargs)  # [..., channel, ...]
  File "/Users/user/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/Users/user/sci/se3cnn/se3cnn/point/operations.py", line 19, in forward
    k = self.kernel(rb - ra)  # [batch, a, b, i, j]
  File "/Users/user/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/Users/user/sci/se3cnn/se3cnn/point/kernel.py", line 120, in forward
    Ys = self.sh(self.set_of_l_filters, r)  # [l_filter * m_filter, batch]
  File "/Users/user/sci/se3cnn/se3cnn/SO3.py", line 304, in spherical_harmonics_xyz
    out = spherical_harmonics(order, alpha, beta)  # [m, ...]
  File "/Users/user/sci/se3cnn/se3cnn/SO3.py", line 274, in spherical_harmonics
    Y = sh(Js, Ms, math.pi - beta.cpu().numpy(), alpha.cpu().numpy())
RuntimeError: Can't call numpy() on Variable that requires grad. Use var.detach().numpy() instead.
```

bkmi committed:
removed specific error message

bkmi commented:
> for me it looks like `examples/point/equivariance.py` should be put into the test files and use the same format as the other files in this directory.

I though the purpose of this file was so that a user would have a minimum example to learn how to test equivariance. I'm happy to put it in test but I think there must be some kind of note in the readme letting the user know about it. I would never read the tests as a user. I also think they serve slightly different purposes since the `examples/point/equivariance.py` code is commented.

In fact, a lot of this information is more-or-less in the tests already. I noticed a normal GatedBlock test is missing, but the Kernel is tested for rotation and parity already.

bkmi committed:
added equivariance tests to tests

mariogeiger commented:
Ok I agree it makes sense
